x-pack/plugins/infra/public/pages/logs/log_entry_rate/sections/anomalies/table.tsx

```diff
@@ -32,73 +34,122 @@ interface SortingOptions {
+    defaultMessage: 'Less log messages in this dataset than expected',
   }
 );
 
+const getAnomalyMessage = (actualRate: number, typicalRate: number): string => {
+  return actualRate < typicalRate ? lessThanExpectedAnomalyMessage : moreThanExpectedAnomalyMessage;
+};
+
 export const AnomaliesTable: React.FunctionComponent<{
   results: LogEntryRateResults;
   setTimeRange: (timeRange: TimeRange) => void;
   timeRange: TimeRange;
   jobId: string;
 }> = ({ results, timeRange, setTimeRange, jobId }) => {
   const tableItems: TableItem[] = useMemo(() => {
-    return Object.entries(results.partitionBuckets).map(([key, value]) => {
+    return results.anomalies.map((anomaly) => {
       return {
+        id: anomaly.id,
         // The real ID
-        partitionId: key,
+        dataset: anomaly.partitionId,
         // Note: EUI's table expanded rows won't work with a key of '' in itemIdToExpandedRowMap, so we have to use the friendly name here
-        partitionName: getFriendlyNameForPartitionId(key),
-        topAnomalyScore: formatAnomalyScore(value.topAnomalyScore),
+        datasetName: getFriendlyNameForPartitionId(anomaly.partitionId),
+        anomalyScore: formatAnomalyScore(anomaly.anomalyScore),
+        anomalyMessage: getAnomalyMessage(anomaly.actualLogEntryRate, anomaly.typicalLogEntryRate),
       };
     });
   }, [results]);
 
-  const [expandedDatasetIds, { add: expandDataset, remove: collapseDataset }] = useSet<string>(
-    new Set()
-  );
+  const [expandedIds, { add: expandDataset, remove: collapseDataset }] = useSet<string>(new Set());
 
   const expandedDatasetRowContents = useMemo(
     () =>
-      [...expandedDatasetIds].reduce<Record<string, React.ReactNode>>(
-        (aggregatedDatasetRows, datasetId) => {
-          return {
-            ...aggregatedDatasetRows,
-            [getFriendlyNameForPartitionId(datasetId)]: (
-              <AnomaliesTableExpandedRow
-                partitionId={datasetId}
-                results={results}
-                setTimeRange={setTimeRange}
-                timeRange={timeRange}
-                jobId={jobId}
-              />
-            ),
-          };
-        },
-        {}
-      ),
-    [expandedDatasetIds, jobId, results, setTimeRange, timeRange]
+      [...expandedIds].reduce<Record<string, React.ReactNode>>((aggregatedDatasetRows, id) => {
+        const anomaly = results.anomalies.find((_anomaly) => _anomaly.id === id);
+
+        return {
+          ...aggregatedDatasetRows,
+          [id]: anomaly ? (
+            <AnomaliesTableExpandedRow anomaly={anomaly} timeRange={timeRange} jobId={jobId} />
+          ) : null,
+        };
+      }, {}),
+    [expandedIds, results, timeRange, jobId]
   );
 
   const [sorting, setSorting] = useState<SortingOptions>({
     sort: {
-      field: 'topAnomalyScore',
+      field: 'anomalyScore',
       direction: 'desc',
     },
   });
 
+  const [_pagination, setPagination] = useState<PaginationOptions>({
```

Comment:
Do we plan on performing the pagination on the HTTP API level when expanding to include other anomaly types?

Reply:
I'm about to launch in to a ramble so I apologise 😬

On the other branch I'm working on I currently have this set so that the API returns a maximum of 1,000 anomaly entries, and whilst they're not rendered yet I intended to hand this off to client side pagination still. 

In my opinion people aren't going to use a view like this to navigate between more than 1,000 entries, in fact I'd be surprised if people navigate past page 2 or 3. My assumption is people will filter further with a narrower time range, or use the dataset filtering (which isn't added yet, but is on the designs).

I remember this discussion coming up in either the Kibana or EUI channels a couple of months back, and the guidance was to show something like "showing up to NUMBER_OF_RESULTS results".

I'm not opposed to adding the pagination to the API, but as with a lot of table based pagination scenarios there's no perfect solution, at least not with my assumptions, which are:

1) We do some pagination via composite aggregations and `after_key`s in logs, but this is normally in the sense of using them to gather "all" of something server side to then send back to the client. Regardless, it'd be irrelevant as aggregations aren't used for the new anomaly API.

2) To do number based pagination (1, 2, 3, 4...) we'd need to know the total number of entries, and therefore total number of pages, up front (could be very costly). And there isn't an elegant way of moving through entries in this way built into Elasticsearch; `offset` and `limit` style (or `from` and `size` for ES) always has the caveat that it becomes more costly the further you paginate. I [see from the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#request-body-search-search-after) that `index.max_result_window` is set to 10,000 by default.

3) Scroll API definitely isn't suitable: `Scroll api is recommended for efficient deep scrolling but scroll contexts are costly and it is not recommended to use it for real time user requests.`.

4) ES handles infinite scroll / stream / load more style pagination just fine with `search_after` and a cursor, but that would only allow "next" and "previous" style pagination within a table. And, as far as I understand, the way most people implement "previous" is to reverse the sort and still use `search_after` with the cursor. Certainly wouldn't suit page based pagination: `search_after is not a solution to jump freely to a random page but rather to scroll many queries in parallel.`.

Given these caveats, if the pagination moves to the server side, I think it'd have to be using `Previous | Next` style pagination to work with the table. And at that point I wonder what the real benefit is over passing 1,000 entries (or perhaps even a smaller number) to the client side? Fetching 1,000 shouldn't be too costly for ES, as it's a relatively simple filter limited query with no complex aggregations.

I'd love to hear your opinion as it's possible there are solutions I don't know about 😄

Reply:
I think your arguments are sound. The `search_after` argument on the query level indeed seems like the only option to paginate in a scalable way at the cost of knowing the page count in advance. On the other hand I don't know if just having Previous/Next pagination is a downside. What is the use-case for jumping to page 5 given that there's no way for the user to know in advance what will be on that page.

As far as API design goes, supporting a pagination cursor and leaving the count arguments up to the UI would relieve us from the burden of having to hard-code a limit. And as we see from the observability home page effort such an API will grow to new use-cases, which would have to be made aware of such assumed limits as well.

So unless the complexity increases significantly I'd still vote for the approach of using the page size and prev/next pagination in the future anomaly-centric API. I understood you're already in the process of implementing that? Can you already judge the additional effort this would require?

Reply:
> On the other hand I don't know if just having Previous/Next pagination is a downside. What is the use-case for jumping to page 5 given that there's no way for the user to know in advance what will be on that page.

Yep, fair point. Like I said before I really don't see people navigating between many pages before filtering further (or going somewhere else - stream, ML etc).

> As far as API design goes, supporting a pagination cursor and leaving the count arguments up to the UI would relieve us from the burden of having to hard-code a limit.

Yep 👍

> I understood you're already in the process of implementing that? Can you already judge the additional effort this would require?

Yep. The API was "done" insofar as supporting what it needed to (dataset filtering would have been one of the next bits). I'll add this to my list to add in next over on that work. Shouldn't be much additional effort at all, certainly not from the API side. Off the top of my head I don't think EUI tables support "next / previous" style pagination via the baked in pagination options, so might need a small custom bit of code there 🤔 

Will add this over on the other branch 👌